### PR TITLE
Improve typing for under_cached_property decorator

### PR DESCRIPTION
**Pull request by bdraco**

### CODE REVIEW

File: src/propcache/_helpers_py.py [hunk 1]
```diff
@@ -27,18 +45,27 @@ def __init__(self, wrapped: Callable[..., _T]) -> None:
         self.__doc__ = wrapped.__doc__
         self.name = wrapped.__name__
 
-    def __get__(self, inst: _TSelf[_T], owner: Optional[Type[Any]] = None) -> _T:
+    @overload
+    def __get__(  # pragma: no cover
+        self, inst: None, owner: Optional[Type[Any]] = None
```
Comment (webknjaz) [inline src/propcache/_helpers_py.py:50]: FTR, `Any` here is what makes MyPy think it's not fully covered.

File: src/propcache/_helpers_py.py [hunk 2]
```diff
@@ -27,18 +45,27 @@ def __init__(self, wrapped: Callable[..., _T]) -> None:
         self.__doc__ = wrapped.__doc__
         self.name = wrapped.__name__
 
-    def __get__(self, inst: _TSelf[_T], owner: Optional[Type[Any]] = None) -> _T:
+    @overload
+    def __get__(  # pragma: no cover
```
Comment (webknjaz) [inline src/propcache/_helpers_py.py:49]: @bdraco why did you add the “no cover” pragma here? It only influences Coveragepy, not MyPy. Does Coveragepy not hit this line? I'm pretty sure it should because the declaration is always executed in runtime…

Now, Coveragepy marks the line as “unmeasured” and MyPy's “partial” coverage overlays that fully, since Coveragepy does not contribute anything to the metric on this line.

Comment (bdraco) [reply src/propcache/_helpers_py.py:49]: Coveragepy doesn’t hit on overloads since they are never executed at run time

File: src/propcache/_helpers_py.py (hunk 1 above)
Comment (bdraco) [reply src/propcache/_helpers_py.py:50]: That’s unfortunate as Any is correct here (also similar to the standard lib functools cachedproperty type shead)

File: src/propcache/_helpers_py.py (hunk 2 above)
Comment (webknjaz) [reply src/propcache/_helpers_py.py:49]: I thought this was about the function declarations. Have you seen any Coveragepy guidance on this?

File: src/propcache/_helpers_py.py (hunk 1 above)
Comment (Dreamsorcerer) [reply src/propcache/_helpers_py.py:50]: Is it? The parameter isn't even used, so `object` would be type safe. Any just disables type checking.

File: src/propcache/_helpers_py.py (hunk 2 above)
Comment (Dreamsorcerer) [reply src/propcache/_helpers_py.py:49]: The declaration should still be executed. Then we usually have coveragerc config to ignore the `...`.

Comment (webknjaz) [reply src/propcache/_helpers_py.py:49]: @Dreamsorcerer yeah, my understanding is that `covdefaults` already excludes `...`.

@bdraco the concern here is that the declaration lines are excluded, and they shouldn't be. It's the body that isn't hit in runtime. But it should already be excluded automatically by the configuration we have in place.

File: src/propcache/_helpers_py.py (hunk 1 above)
Comment (webknjaz) [reply src/propcache/_helpers_py.py:50]: Yeah, I had the same thought. Just haven't had time to look into what should be put in here instead. I recently realized that such undercovered things also influence how coverage is computed in the calling projects, so it's quite important to improve the precision here.

File: src/propcache/_helpers_py.py (hunk 2 above)
Comment (bdraco) [reply src/propcache/_helpers_py.py:49]: Removal attempt in https://github.com/aio-libs/propcache/pull/48

File: src/propcache/_helpers_py.py (hunk 1 above)
Comment (bdraco) [reply src/propcache/_helpers_py.py:50]: https://github.com/python/typeshed/blob/dbd0d3521745288a3b2e345d8683bb9539d28d60/stdlib/functools.pyi#L192

Here is what standard lib uses for `cached_property`

Comment (Dreamsorcerer) [reply src/propcache/_helpers_py.py:50]: If the implementation is type safe with anything, then it should be updated to object. I've had discussions in typeshed before, and the consensus is that typeshed should use object in such circumstances (otherwise, it triggers Any warnings in mypy if all the any warnings are enabled, anytime a user uses that function).

Comment (bdraco) [reply src/propcache/_helpers_py.py:50]: attempt here https://github.com/aio-libs/propcache/pull/50

Comment (webknjaz) [reply src/propcache/_helpers_py.py:50]: @bdraco some of those types are historic due to limitations that existed in the past: https://github.com/python/mypy/issues/11878#issuecomment-1147925317.
